staging/src/k8s.io/csi-translation-lib/plugins/rbd.go

```diff
@@ -0,0 +1,313 @@
+	if pv.Spec.RBD.ReadOnly {
+		am = v1.ReadOnlyMany
+	} else {
+		am = v1.ReadWriteOnce
+	}
+	pv.Spec.AccessModes = []v1.PersistentVolumeAccessMode{am}
+	csiSource := &v1.CSIPersistentVolumeSource{
+		Driver:                    RbdDriverName,
+		FSType:                    pv.Spec.RBD.FSType,
+		VolumeHandle:              volID,
+		VolumeAttributes:          volumeAttributes,
+		NodeStageSecretRef:        pv.Spec.RBD.SecretRef,
+		ControllerExpandSecretRef: pv.Spec.RBD.SecretRef,
+	}
+	pv.Spec.RBD = nil
+	pv.Spec.CSI = csiSource
+	return pv, nil
+}
+
+// TranslateCSIPVToInTree takes a PV with a CSI PersistentVolume Source and will translate
+// it to an in-tree Persistent Volume Source for the in-tree volume
+func (p rbdCSITranslator) TranslateCSIPVToInTree(pv *v1.PersistentVolume) (*v1.PersistentVolume, error) {
+	if pv == nil || pv.Spec.CSI == nil {
+		return nil, fmt.Errorf("pv is nil or CSI source not defined on pv")
+	}
+	var rbdImageName string
+	var monSlice []string
+	csiSource := pv.Spec.CSI
+
+	rbdImageName = csiSource.VolumeAttributes[imgNameKey]
+	rbdPool := csiSource.VolumeAttributes[poolKey]
+	radosUser := csiSource.VolumeAttributes[adminIDKey]
+	if radosUser == "" {
+		radosUser = defaultAdminUser
+	}
+
+	RBDSource := &v1.RBDPersistentVolumeSource{
+		CephMonitors: monSlice,
+		RBDImage:     rbdImageName,
+		FSType:       csiSource.FSType,
+		RBDPool:      rbdPool,
+		RadosUser:    radosUser,
+		ReadOnly:     csiSource.ReadOnly,
+	}
+
+	if pv.Annotations == nil {
+		pv.Annotations = make(map[string]string)
+	}
+	fillAnnotationsFromCSISource(pv, csiSource)
+	nodeSecret := csiSource.NodeStageSecretRef
+	if nodeSecret != nil {
+		RBDSource.SecretRef = &v1.SecretReference{Name: nodeSecret.Name, Namespace: nodeSecret.Namespace}
+	}
+	pv.Spec.CSI = nil
+	pv.Spec.RBD = RBDSource
+
+	return pv, nil
+}
+
+// CanSupport tests whether the plugin supports a given persistent volume
+// specification from the API.
+func (p rbdCSITranslator) CanSupport(pv *v1.PersistentVolume) bool {
+	return pv != nil && pv.Spec.RBD != nil
+}
+
+// CanSupportInline tests whether the plugin supports a given inline volume
+// specification from the API.
+func (p rbdCSITranslator) CanSupportInline(volume *v1.Volume) bool {
+	return volume != nil && volume.RBD != nil
+}
+
+// GetInTreePluginName returns the in-tree plugin name this migrates
+func (p rbdCSITranslator) GetInTreePluginName() string {
+	return RbdVolumePluginName
+}
+
+// GetCSIPluginName returns the name of the CSI plugin that supersedes the in-tree plugin
+func (p rbdCSITranslator) GetCSIPluginName() string {
+	return RbdDriverName
+}
+
+// RepairVolumeHandle generates a correct volume handle based on node ID information.
+func (p rbdCSITranslator) RepairVolumeHandle(volumeHandle, nodeID string) (string, error) {
+	return volumeHandle, nil
+}
+
+// fillDefaultSCParams fills some sc parameters with default values
+func fillDefaultSCParams(params map[string]string) {
```

Comment:
Are these default values required for CSI but not required for intree?

Reply:
@msau42 yes it is required for csi and also intree fills a couple of  default values like pool which is handled here.